File: src/game/player-helpers.js
// ── Opponent helper ─────────────────────────────────────────────────────────

export function opponentPlayerNum(pn) { return pn === 1 ? 2 : 1; }

// ── Initiative helper ───────────────────────────────────────────────────────

export function getInitiativePlayerNum(game) {
  return game.initiativePlayerId === game.player1Id ? 1 : 2;
}

// Player-number property accessors — eliminates `pn === 1 ? game.p1X : game.p2X` ternaries.

// ── Getters (read-only access) ──────────────────────────────────────────────

export function getPlayerId(game, pn)             { return pn === 1 ? game.player1Id : game.player2Id; }
export function getDcList(game, pn)               { return pn === 1 ? game.p1DcList : game.p2DcList; }
export function getDcMessageIds(game, pn)          { return pn === 1 ? game.p1DcMessageIds : game.p2DcMessageIds; }
export function getHandChannelId(game, pn)         { return pn === 1 ? game.p1HandId : game.p2HandId; }
export function getPlayAreaId(game, pn)            { return pn === 1 ? game.p1PlayAreaId : game.p2PlayAreaId; }
export function getActivationsRemaining(game, pn)  { return pn === 1 ? game.p1ActivationsRemaining : game.p2ActivationsRemaining; }
export function getActivationsTotal(game, pn)      { return pn === 1 ? game.p1ActivationsTotal : game.p2ActivationsTotal; }
export function getActivationsMessageId(game, pn)  { return pn === 1 ? game.p1ActivationsMessageId : game.p2ActivationsMessageId; }
export function getActivatedDcIndices(game, pn)    { return pn === 1 ? game.p1ActivatedDcIndices : game.p2ActivatedDcIndices; }
export function getDiscardThreadId(game, pn)       { return pn === 1 ? game.p1DiscardThreadId : game.p2DiscardThreadId; }
export function getSquad(game, pn)                 { return pn === 1 ? game.player1Squad : game.player2Squad; }
export function getCcHand(game, pn)                { return pn === 1 ? game.player1CcHand : game.player2CcHand; }
export function getCcDiscard(game, pn)             { return pn === 1 ? game.player1CcDiscard : game.player2CcDiscard; }
export function getCcDeck(game, pn)                { return pn === 1 ? game.player1CcDeck : game.player2CcDeck; }
export function getCcAttachments(game, pn)         { return pn === 1 ? game.p1CcAttachments : game.p2CcAttachments; }
export function getDcAttachments(game, pn)         { return pn === 1 ? game.p1DcAttachments : game.p2DcAttachments; }

// ── Health state sync ────────────────────────────────────────────────────────

/** Sync a figure's health state array back to the DC list for persistence. */
export function syncHealthStateToList(game, playerNum, msgId, healthState) {
  const dcIds = getDcMessageIds(game, playerNum);
  const dcList = getDcList(game, playerNum);
  const idx = dcIds ? dcIds.indexOf(msgId) : -1;
  if (idx >= 0 && dcList?.[idx]) dcList[idx].healthState = [...healthState];
}

// ── Setters (for properties that are reassigned, not just mutated) ──────────

export function setActivationsRemaining(game, pn, v) { if (pn === 1) game.p1ActivationsRemaining = v; else game.p2ActivationsRemaining = v; }
export function setActivationsTotal(game, pn, v)     { if (pn === 1) game.p1ActivationsTotal = v; else game.p2ActivationsTotal = v; }
export function setActivatedDcIndices(game, pn, v)   { if (pn === 1) game.p1ActivatedDcIndices = v; else game.p2ActivatedDcIndices = v; }

// ── Activation count recompute (single source of truth) ─────────────────────

/**
 * Recompute ActivationsTotal and ActivationsRemaining from board state.
 * A DC is activatable if it has figures on the board (alive and deployed).
 * Lie-in-Ambush set-aside groups are "out of play" and do NOT count until deployed.
 * Remaining = activatable DCs that haven't been activated this round.
 *
 * Call this instead of hand-rolling ±1 delta math on activation counts.
 * @returns {{ total: number, remaining: number }}
 */
export function recomputeActivationCounts(game, pn) {
  const dcList = getDcList(game, pn) || [];
  const figs = game.figurePositions?.[pn] || {};
  const figKeys = Object.keys(figs);
  const activatedIndices = getActivatedDcIndices(game, pn) || [];

  let total = 0;
  let activated = 0;

  for (let i = 0; i < dcList.length; i++) {
    const dcName = dcList[i]?.dcName || dcList[i];
    // Skip figureless DCs (skirmish upgrades like [Extra Armor])
    if (/^\[.+\]$/.test(dcName)) continue;

    const hasFigures = figKeys.some(fk => fk.startsWith(dcName + '-') && figs[fk]);

    if (hasFigures) {
      total++;
      if (activatedIndices.includes(i)) activated++;
    }
  }

  setActivationsTotal(game, pn, total);
  setActivationsRemaining(game, pn, total - activated);
  return { total, remaining: total - activated };
}

// ── Mutations ────────────────────────────────────────────────────────────────

/** Remove a figure's position and clean up per-figure state (device tokens, conditions). */
export function removeFigurePosition(game, pn, figureKey) {
  if (game.figurePositions?.[pn]) delete game.figurePositions[pn][figureKey];
  if (game.deviceTokens?.[figureKey]) delete game.deviceTokens[figureKey];
  if (game.figureConditions?.[figureKey]) delete game.figureConditions[figureKey];
}

/**
 * Canonical push/pull/displacement write. Moves a figure that is already on the
 * board to a new space. Returns { prevPos, newPos } or null if the figure has
 * no current position (guard — prevents ghost writes).
 */
export function pushFigure(game, playerNum, figureKey, newSpace) {
  const positions = game.figurePositions?.[playerNum];
  if (!positions) return null;
  const prevPos = positions[figureKey];
  if (prevPos == null) return null;
  positions[figureKey] = String(newSpace).toLowerCase();
  return { prevPos, newPos: positions[figureKey] };
}

// ── Key helpers (for code that needs both read + write via game[key]) ───────

export function ccHandKey(pn)       { return pn === 1 ? 'player1CcHand' : 'player2CcHand'; }
export function ccDiscardKey(pn)    { return pn === 1 ? 'player1CcDiscard' : 'player2CcDiscard'; }
export function ccDeckKey(pn)       { return pn === 1 ? 'player1CcDeck' : 'player2CcDeck'; }
export function ccDrawnKey(pn)      { return pn === 1 ? 'player1CcDrawn' : 'player2CcDrawn'; }
export function ccAttachmentsKey(pn) { return pn === 1 ? 'p1CcAttachments' : 'p2CcAttachments'; }
export function dcAttachmentsKey(pn) { return pn === 1 ? 'p1DcAttachments' : 'p2DcAttachments'; }
export function dcAttachmentMessageIdsKey(pn) { return pn === 1 ? 'p1DcAttachmentMessageIds' : 'p2DcAttachmentMessageIds'; }
export function vpKey(pn)           { return pn === 1 ? 'player1VP' : 'player2VP'; }
export function deployMetadataKey(pn) { return pn === 1 ? 'player1DeployMetadata' : 'player2DeployMetadata'; }
export function deployLabelsKey(pn)   { return pn === 1 ? 'player1DeployLabels' : 'player2DeployLabels'; }
export function armyCostModifierKey(pn) { return pn === 1 ? 'player1ArmyCostModifier' : 'player2ArmyCostModifier'; }
export function activatedDcIndicesKey(pn) { return pn === 1 ? 'p1ActivatedDcIndices' : 'p2ActivatedDcIndices'; }

// ── DC / playableBy matching ────────────────────────────────────────────────

/**
 * Check whether a DC name/keywords match a playableBy restriction.
 * Shared by headless CC play and Discord CC hand attachment filter.
 * @param {string} dcName - DC name (may include [DG N] suffix)
 * @param {string} playableBy - playableBy string from CC effect
 * @param {Function} getDcEffectsFn - returns all DC effects object
 * @param {Function} getDcKeywordsFn - returns keywords map (game => {dcName: [kw...]})
 * @param {object} game - game state
 * @param {string} [displayName] - optional display name for name matching (Discord only)
 * @returns {boolean}
 */
export function dcMatchesPlayableBy(dcName, playableBy, getDcEffectsFn, getDcKeywordsFn, game, displayName) {
  if (!playableBy) return true;
  const lower = playableBy.toLowerCase().trim();
  if (!lower || lower === 'any figure') return true;

  const dcBase = dcName.replace(/\s*\[(?:DG|Group) \d+\]$/i, '').replace(/\s*\((?:Elite|Regular)\)\s*$/i, '').trim();
  const allDcEffects = (getDcEffectsFn ? getDcEffectsFn() : null) || {};
  const dcData = allDcEffects[dcName] || allDcEffects[dcBase] || {};
  const kwRaw = getDcKeywordsFn
    ? (getDcKeywordsFn(game)?.[dcName] || getDcKeywordsFn(game)?.[dcBase] || [])
    : (dcData.keywords || []);
  const kwLower = kwRaw.map(k => String(k).toLowerCase());
  const affiliationLower = (dcData.affiliation || '').toLowerCase();

  const AFFILIATIONS = new Set(['imperial', 'rebel', 'scum', 'mercenary']);
  const alternatives = lower.split(/\s+or\s+/i).map(a => a.trim().replace(/^"|"$/g, ''));

  for (const alt of alternatives) {
    if (alt === 'unique' || alt === 'any unique figure') {
      if (dcData.unique) return true;
      continue;
    }
    if (alt === 'any small figure') {
      if (kwLower.includes('small')) return true;
      continue;
    }
    // Name match — check both dcName and optional displayName
    const dcLow = dcBase.toLowerCase();
    if (dcLow.includes(alt) || alt.includes(dcLow)) return true;
    if (displayName) {
      const dispBase = String(displayName)
        .replace(/\s*\[(?:DG|Group) \d+\]$/i, '')
        .replace(/\s*\((?:Elite|Regular)\)\s*$/i, '')
        .trim().toLowerCase();
      if (dispBase.includes(alt) || alt.includes(dispBase)) return true;
    }
    // Decompose into affiliation + keyword parts
    const words = alt.split(/\s+/);
    let reqAff = null;
    const reqKwWords = [];
    for (const w of words) {
      if (AFFILIATIONS.has(w) && !reqAff) reqAff = w;
      else reqKwWords.push(w);
    }
    const reqKw = reqKwWords.join(' ');
    if (reqAff && affiliationLower !== reqAff && affiliationLower !== 'any') continue;
    if (reqKw && !kwLower.includes(reqKw)) continue;
    if (reqAff || reqKw) return true;
  }
  return false;
}

File: python/engine/mechanics/player_helpers.py
"""Per-player getters/setters + squad helpers — mirror of src/game/player-helpers.js.

Almost all functions are 2-branch ternaries (p1X vs p2X); the heavier
logic lives in recompute_activation_counts and dc_matches_playable_by.

Field names stay JS-native (camelCase) during the parity window because
the game dict is still written by JS during cross-validation. After
Phase 14 (JS deletion) callers can either keep camelCase or rewrap.
"""
from __future__ import annotations

import re
from typing import Any, Callable, Dict, List, Optional


# ---------------------------------------------------------------------------
# Opponent / initiative

def opponent_player_num(pn: int) -> int:
    return 2 if pn == 1 else 1


def get_initiative_player_num(game: Any) -> int:
    data = _data(game)
    return 1 if data.get('initiativePlayerId') == data.get('player1Id') else 2


# ---------------------------------------------------------------------------
# Squad cost

def sum_squad_dc_cost(squad: Optional[Dict[str, Any]],
                      stats_lookup: Callable[[str], Optional[Dict[str, Any]]]) -> int:
    """Sum DC costs for a squad. Handles bracketed attachments + Regular/Elite variants."""
    dc_list = (squad or {}).get('dcList') or []
    total = 0
    for entry in dc_list:
        name = entry.strip('[]') if isinstance(entry, str) else entry
        stats = (
            stats_lookup(name)
            or stats_lookup(f'[{name}]')
            or stats_lookup(f'{name} (Regular)')
            or stats_lookup(f'{name} (Elite)')
        )
        if stats and stats.get('cost') is not None:
            total += stats['cost']
    return total


# ---------------------------------------------------------------------------
# Getters (pn → field)

def _get(game: Any, p1_key: str, p2_key: str, pn: int) -> Any:
    data = _data(game)
    return data.get(p1_key) if pn == 1 else data.get(p2_key)


def get_player_id(game, pn):                 return _get(game, 'player1Id',               'player2Id',               pn)
def get_dc_list(game, pn):                   return _get(game, 'p1DcList',                'p2DcList',                pn)
def get_dc_message_ids(game, pn):            return _get(game, 'p1DcMessageIds',          'p2DcMessageIds',          pn)
def get_hand_channel_id(game, pn):           return _get(game, 'p1HandId',                'p2HandId',                pn)
def get_play_area_id(game, pn):              return _get(game, 'p1PlayAreaId',            'p2PlayAreaId',            pn)
def get_activations_remaining(game, pn):     return _get(game, 'p1ActivationsRemaining',  'p2ActivationsRemaining',  pn)
def get_activations_total(game, pn):         return _get(game, 'p1ActivationsTotal',      'p2ActivationsTotal',      pn)
def get_activations_message_id(game, pn):    return _get(game, 'p1ActivationsMessageId',  'p2ActivationsMessageId',  pn)
def get_activated_dc_indices(game, pn):      return _get(game, 'p1ActivatedDcIndices',    'p2ActivatedDcIndices',    pn)
def get_discard_thread_id(game, pn):         return _get(game, 'p1DiscardThreadId',       'p2DiscardThreadId',       pn)
def get_squad(game, pn):                     return _get(game, 'player1Squad',            'player2Squad',            pn)
def get_cc_hand(game, pn):                   return _get(game, 'player1CcHand',           'player2CcHand',           pn)
def get_cc_discard(game, pn):                return _get(game, 'player1CcDiscard',        'player2CcDiscard',        pn)
def get_cc_deck(game, pn):                   return _get(game, 'player1CcDeck',           'player2CcDeck',           pn)
def get_cc_attachments(game, pn):            return _get(game, 'p1CcAttachments',         'p2CcAttachments',         pn)
def get_dc_attachments(game, pn):            return _get(game, 'p1DcAttachments',         'p2DcAttachments',         pn)


# ---------------------------------------------------------------------------
# Health state sync

def sync_health_state_to_list(game: Any, player_num: int, msg_id: str,
                              health_state: List) -> None:
    """Sync a figure's health state array back to the DC list for persistence."""
    dc_ids = get_dc_message_ids(game, player_num) or []
    dc_list = get_dc_list(game, player_num) or []
    if msg_id not in dc_ids:
        return
    idx = dc_ids.index(msg_id)
    if idx >= 0 and idx < len(dc_list) and isinstance(dc_list[idx], dict):
        dc_list[idx]['healthState'] = list(health_state)


# ---------------------------------------------------------------------------
# Setters

def _set(game: Any, p1_key: str, p2_key: str, pn: int, v: Any) -> None:
    data = _data(game)
    data[p1_key if pn == 1 else p2_key] = v


def set_activations_remaining(game, pn, v):  _set(game, 'p1ActivationsRemaining', 'p2ActivationsRemaining', pn, v)
def set_activations_total(game, pn, v):      _set(game, 'p1ActivationsTotal',     'p2ActivationsTotal',     pn, v)
def set_activated_dc_indices(game, pn, v):   _set(game, 'p1ActivatedDcIndices',   'p2ActivatedDcIndices',   pn, v)


# ---------------------------------------------------------------------------
# Activation count recompute (single source of truth)

_SKIRMISH_UPGRADE_RE = re.compile(r'^\[.+\]$')


def recompute_activation_counts(game: Any, pn: int) -> Dict[str, int]:
    """Recompute ActivationsTotal / ActivationsRemaining from board state.

    A DC is activatable if it has figures on the board. Lie-in-Ambush
    set-aside groups are out-of-play and do NOT count until deployed.
    Mirrors JS recomputeActivationCounts.
    """
    data = _data(game)
    dc_list = get_dc_list(game, pn) or []
    figs = (data.get('figurePositions') or {}).get(pn) or {}
    fig_keys = list(figs.keys())
    activated_indices = get_activated_dc_indices(game, pn) or []

    total = 0
    activated = 0
    dg_counts: Dict[str, int] = {}

    for i, entry in enumerate(dc_list):
        dc_name = entry.get('dcName') if isinstance(entry, dict) else entry
        if not isinstance(dc_name, str):
            continue
        if _SKIRMISH_UPGRADE_RE.match(dc_name):
            continue
        dg_counts[dc_name] = dg_counts.get(dc_name, 0) + 1
        dg_index = dg_counts[dc_name]
        prefix = f'{dc_name}-{dg_index}-'
        has_figures = any(fk.startswith(prefix) and figs.get(fk) for fk in fig_keys)
        if has_figures:
            total += 1
            if i in activated_indices:
                activated += 1

    set_activations_total(game, pn, total)
    set_activations_remaining(game, pn, total - activated)
    return {'total': total, 'remaining': total - activated}


# ---------------------------------------------------------------------------
# Figure position mutations

def remove_figure_position(game: Any, pn: int, figure_key: str) -> None:
    """Remove a figure's position and clean up per-figure state."""
    data = _data(game)
    positions = (data.get('figurePositions') or {}).get(pn)
    if isinstance(positions, dict) and figure_key in positions:
        del positions[figure_key]
    device_tokens = data.get('deviceTokens')
    if isinstance(device_tokens, dict) and figure_key in device_tokens:
        del device_tokens[figure_key]
    conditions = data.get('figureConditions')
    if isinstance(conditions, dict) and figure_key in conditions:
        del conditions[figure_key]


def push_figure(game: Any, player_num: int, figure_key: str,
                new_space: str) -> Optional[Dict[str, str]]:
    """Move a figure already on the board to a new space.

    Returns {'prevPos', 'newPos'} or None if figure has no current position.
    """
    data = _data(game)
    positions = (data.get('figurePositions') or {}).get(player_num)
    if not isinstance(positions, dict):
        return None
    prev_pos = positions.get(figure_key)
    if prev_pos is None:
        return None
    new_pos = str(new_space).lower()
    positions[figure_key] = new_pos
    return {'prevPos': prev_pos, 'newPos': new_pos}


# ---------------------------------------------------------------------------
# Key helpers (for code that needs both read+write via game[key])

def cc_hand_key(pn):                    return 'player1CcHand' if pn == 1 else 'player2CcHand'
def cc_discard_key(pn):                 return 'player1CcDiscard' if pn == 1 else 'player2CcDiscard'
def cc_deck_key(pn):                    return 'player1CcDeck' if pn == 1 else 'player2CcDeck'
def cc_drawn_key(pn):                   return 'player1CcDrawn' if pn == 1 else 'player2CcDrawn'
def cc_attachments_key(pn):             return 'p1CcAttachments' if pn == 1 else 'p2CcAttachments'
def dc_attachments_key(pn):             return 'p1DcAttachments' if pn == 1 else 'p2DcAttachments'
def dc_attachment_message_ids_key(pn):  return 'p1DcAttachmentMessageIds' if pn == 1 else 'p2DcAttachmentMessageIds'
def vp_key(pn):                         return 'player1VP' if pn == 1 else 'player2VP'
def deploy_metadata_key(pn):            return 'player1DeployMetadata' if pn == 1 else 'player2DeployMetadata'
def deploy_labels_key(pn):              return 'player1DeployLabels' if pn == 1 else 'player2DeployLabels'
def army_cost_modifier_key(pn):         return 'player1ArmyCostModifier' if pn == 1 else 'player2ArmyCostModifier'
def activated_dc_indices_key(pn):       return 'p1ActivatedDcIndices' if pn == 1 else 'p2ActivatedDcIndices'


# ---------------------------------------------------------------------------
# DC / playableBy matching

_DG_SUFFIX_RE = re.compile(r'\s*\[(?:DG|Group) \d+\]$', re.IGNORECASE)
_VARIANT_SUFFIX_RE = re.compile(r'\s*\((?:Elite|Regular)\)\s*$', re.IGNORECASE)

_AFFILIATIONS = frozenset({'imperial', 'rebel', 'scum', 'mercenary'})


def dc_matches_playable_by(dc_name: str,
                           playable_by: Optional[str],
                           get_dc_effects_fn: Optional[Callable[[], Dict[str, Any]]],
                           get_dc_keywords_fn: Optional[Callable[[Any], Dict[str, List[str]]]],
                           game: Any,
                           display_name: Optional[str] = None) -> bool:
    """Check whether a DC matches a CC's playableBy restriction.

    Mirrors src/game/player-helpers.js:dcMatchesPlayableBy byte-for-byte.
    """
    if not playable_by:
        return True
    lower = playable_by.lower().strip()
    if not lower or lower == 'any figure':
        return True

    dc_base = _VARIANT_SUFFIX_RE.sub('', _DG_SUFFIX_RE.sub('', dc_name)).strip()
    all_dc_effects = (get_dc_effects_fn() if get_dc_effects_fn else None) or {}
    dc_data = all_dc_effects.get(dc_name) or all_dc_effects.get(dc_base) or {}

    if get_dc_keywords_fn:
        kw_map = get_dc_keywords_fn(game) or {}
        kw_raw = kw_map.get(dc_name) or kw_map.get(dc_base) or []
    else:
        kw_raw = dc_data.get('keywords') or []
    kw_lower = [str(k).lower() for k in kw_raw]
    affiliation_lower = str(dc_data.get('affiliation') or '').lower()

    alternatives = [a.strip().strip('"') for a in re.split(r'\s+or\s+', lower, flags=re.IGNORECASE)]

    for alt in alternatives:
        if alt in ('unique', 'any unique figure'):
            if dc_data.get('unique'):
                return True
            continue
        if alt == 'any small figure':
            if 'small' in kw_lower:
                return True
            continue

        dc_low = dc_base.lower()
        if dc_low in alt or alt in dc_low:
            return True
        if display_name:
            disp_base = _VARIANT_SUFFIX_RE.sub('', _DG_SUFFIX_RE.sub('', str(display_name))).strip().lower()
            if disp_base and (disp_base in alt or alt in disp_base):
                return True

        words = alt.split()
        req_aff = None
        req_kw_words: List[str] = []
        for w in words:
            if w in _AFFILIATIONS and not req_aff:
                req_aff = w
            else:
                req_kw_words.append(w)
        req_kw = ' '.join(req_kw_words)
        if req_aff and affiliation_lower not in (req_aff, 'any'):
            continue
        if req_kw and req_kw not in kw_lower:
            continue
        if req_aff or req_kw:
            return True
    return False


# ---------------------------------------------------------------------------

def _data(game: Any) -> Dict[str, Any]:
    """Unwrap GameState → dict if needed. Tolerates plain dicts too."""
    data_attr = getattr(game, 'data', None)
    if isinstance(data_attr, dict):
        return data_attr
    if isinstance(game, dict):
        return game
    raise TypeError(
        f'player_helpers expected GameState or dict, got {type(game).__name__}'
    )
